### PR TITLE
Docs for Gitea Actions

### PR DESCRIPTION
**Pull request by wolfogre**
A new documentation section for Gitea Actions.

Some content comes from:

- [Hacking on Gitea Actions](https://blog.gitea.io/2023/03/hacking-on-gitea-actions/)
- The README of [act_runner](https://gitea.com/gitea/act_runner)
- @ChristopherHX's excellent overview of the differences between Gitea Actions and GitHub Actions in [this comment](https://github.com/go-gitea/gitea/issues/13539#issuecomment-1448888850).

### CODE REVIEW

File: docs/content/doc/actions/act-runner.en-us.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+**Table of Contents**
+
+{{< toc >}}
+
+## Requirements
+
+It is recommended to run jobs in a docker container, so you need to install docker first.
+And make sure that the docker daemon is running.
+
+Other container engines which is compatible with docker API should also work, but it is not tested yet.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]:
```suggestion
Other OCI container engines which are compatible with Docker's API should also work, but are untested.
```

File: docs/content/doc/actions/act-runner.en-us.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+```bash
+docker pull gitea/act_runner:latest # for the latest stable release
+docker pull gitea/act_runner:nightly # for the latest nightly build
+```
+
+## Configuration
+
+The act runner can be configured by a configuration file.
+It is optional, act runner will use the default configuration if you don't specify a configuration file.
+But at least we need to know how to do it when we want to change some behavior of the act runner.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]:
```suggestion
Configuration is done via a configuration file. It is optional, and the default configuration will be used when no configuration file is specified.
```

File: docs/content/doc/actions/act-runner.en-us.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+The default configuration is safe to use without any modification, so you can just use it directly.
+
+```bash
+./act_runner generate-config > config.yaml
+./act_runner --config config.yaml [command]
+```
+
+When you are using the docker image, you can specify the configuration file by using the `CONFIG_FILE` environment variable.
+And please make sure that the configuration file is mounted in the container.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]:
```suggestion
When you are using the docker image, you can specify the configuration file by using the `CONFIG_FILE` environment variable. Make sure that the file is mounted into the container as a volume:
```

File: docs/content/doc/actions/act-runner.en-us.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+### Runner levels
+
+You can register a runner in different levels, it can be:
+
+- Instance level: The runner will run jobs for all repositories in the instance.
+- Organization level: The runner will run jobs for all repositories in the organization.
+- Repository level: The runner will run jobs for the repository it belongs to.
+
+Please note that the repository may still use instance-level or organization-level runners even if it has its own repository-level runners.
+We may provide options to control this in the future.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]:
```suggestion
Note that the repository may still use instance-level or organization-level runners even if it has its own repository-level runners. A future release may provide an option to allow more control over this.
```

File: docs/content/doc/actions/act-runner.en-us.md [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+The level of the runner determines where to obtain the registration token.
+
+- Instance level: The admin settings page, like `<your_gitea.com>/admin/runners`.
+- Organization level: The organization settings page, like `<your_gitea.com>/<org>/settings/runners`.
+- Repository level: The repository settings page, like `<your_gitea.com>/<owner>/<repo>/settings/runners`.
+
+If you cannot see the settings page, please make sure that you have the right permissions and that Actions have been enabled.
+
+The format of the registration token is something like `D0gvfu2iHfUjNqCYVljVyRV14fISpJxxxxxxxxxx`.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]:
```suggestion
The format of the registration token is a random string `D0gvfu2iHfUjNqCYVljVyRV14fISpJxxxxxxxxxx`.
```

File: docs/content/doc/actions/act-runner.en-us.md [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+### Register the runner
+
+The act runner can be registered by running the following command:
+
+```bash
+./act_runner register
+```
+
+Or you can use the `--config` option to specify the configuration file, which is mentioned in the previous section.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]:
```suggestion
Alternatively, you can use the `--config` option to specify the configuration file mentioned in the previous section.

```

File: docs/content/doc/actions/act-runner.en-us.md [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+```bash
+./act_runner --config config.yaml register
+```
+
+And you will be asked to input the registration information step by step. Includes:
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]:
```suggestion
You will be asked to input the registration information step by step. Includes:
```

File: docs/content/doc/actions/act-runner.en-us.md [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+- The Gitea instance URL, like `https://gitea.com/` or `http://192.168.8.8:3000/`.
+- The registration token.
+- The runner name, which is optional. If you leave it blank, the hostname will be used.
+- The runner labels, which is optional. If you leave it blank, the default labels will be used.
+
+You may confuse about the runner labels, we will explain it later.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]:
```suggestion
You may be confused about the runner labels, which will be explained later.
```

File: docs/content/doc/actions/act-runner.en-us.md [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+Please do not edit it manually.
+If this file is missing or corrupted, you can simply remove it and register again.
+
+If you want to store the registration information in another place, you can specify it in the configuration file,
+and don't forget to specify the `--config` option.
+
+### Register the runner with docker
+
+If you are using the docker image, the things are a little different.
+Registration and running are combined into one step, so you need to specify the registration information when running the act runner.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]:
```suggestion
If you are using the docker image, behaviour will be slightly different. Registration and running are combined into one step in this case, so you need to specify the registration information when running the act runner.
```

File: docs/content/doc/actions/act-runner.en-us.md [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+### Labels
+
+The labels of a runner are used to determine which jobs the runner can run, and how to run them.
+
+The default labels are `ubuntu-latest:docker://node:16-bullseye,ubuntu-22.04:docker://node:16-bullseye,ubuntu-20.04:docker://node:16-bullseye,ubuntu-18.04:docker://node:16-buster`.
+It is a comma-separated list, and each item is a label.
+
+Let's take `ubuntu-22.04:docker://node:16-bullseye` as an example.
+It means that the runner can run jobs with `runs-on: ubuntu-22.04`, and the job will be run in a docker container with the image `node:16-bullseye`.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md:183]: `node` images are debian-based, are they really a good example for mapping a ubuntu label to them?

File: docs/content/doc/actions/act-runner.en-us.md [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+If you think the image is too poor, and you have enough disk space to use a better and bigger one, you can change it to `ubuntu-22.04:docker://<the image you like>`.
```
Comment (silverwind) [inline docs/content/doc/actions/act-runner.en-us.md]: What is meant by "poor"?

File: docs/content/doc/actions/comparison.en-us.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+  sidebar:
+    parent: "actions"
+    name: "Comparison"
+    weight: 30
+    identifier: "actions-comparison"
+---
+
+# Compared to GitHub Actions
+
+Although Gitea Actions is designed to be compatible with GitHub Actions, there are still some differences between them.
```
Comment (silverwind) [inline docs/content/doc/actions/comparison.en-us.md]:
```suggestion
Even though Gitea Actions is designed to be compatible with GitHub Actions, there are some differences between them.
```

File: docs/content/doc/actions/comparison.en-us.md [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+**Table of Contents**
+
+{{< toc >}}
+
+## Additional features
+
+### Use absolute actions
+
+Gitea Actions supports using absolute actions, which means that you can use actions from any git repository.
```
Comment (silverwind) [inline docs/content/doc/actions/comparison.en-us.md]:
```suggestion
Gitea Actions supports defining actions via absolute URL, which means that you can use actions from any git repository.
```

File: docs/content/doc/actions/comparison.en-us.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+Although Gitea Actions is designed to be compatible with GitHub Actions, there are still some differences between them.
+
+**Table of Contents**
+
+{{< toc >}}
+
+## Additional features
+
+### Use absolute actions
```
Comment (silverwind) [inline docs/content/doc/actions/comparison.en-us.md]:
```suggestion
### Absolute action URLs
```

File: docs/content/doc/actions/comparison.en-us.md [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+As a workaround, you can use [go-hashfiles](https://gitea.com/actions/go-hashfiles) instead.
+
+## Missing features
+
+### Variables
+
+See [Variables](https://docs.github.com/en/actions/learn-github-actions/variables).
+
+It is under development.
```
Comment (silverwind) [inline docs/content/doc/actions/comparison.en-us.md]:
```suggestion
It's under development.
```

File: docs/content/doc/actions/comparison.en-us.md [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+See [Creating an annotation for an error](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#example-creating-an-annotation-for-an-error)
+
+It's ignored by Gitea Actions now.
+
+## Missing UI features
+
+### Pre and Post steps 
+
+Pre and Post steps don't have their own section in the job log ui.
```
Comment (silverwind) [inline docs/content/doc/actions/comparison.en-us.md]:
```suggestion
Pre and Post steps don't have their own section in the job log user interface.
```

File: docs/content/doc/actions/comparison.en-us.md [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+
+## Different behavior
+
+### Context availability
+
+Context availability is not checked, so you can use the env context on more places.
+See [Context availability](https://docs.github.com/en/actions/learn-github-actions/contexts#context-availability).
+
+## Known issues with some actions
```
Comment (silverwind) [inline docs/content/doc/actions/comparison.en-us.md]:
```suggestion
## Known issues
```

File: docs/content/doc/actions/design.en-us.md [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+  sidebar:
+    parent: "actions"
+    name: "Design"
+    weight: 40
+    identifier: "actions-design"
+---
+
+# Design of Gitea Actions
+
+Gitea Actions has multiple parts.
```
Comment (silverwind) [inline docs/content/doc/actions/design.en-us.md:18]:
```suggestion
Gitea Actions has multiple components.
```

File: docs/content/doc/actions/design.en-us.md [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+Let's explain them one by one.
```
Comment (silverwind) [inline docs/content/doc/actions/design.en-us.md]:
```suggestion
Gitea Actions has multiple components. This document describes them individually.
```

File: docs/content/doc/actions/design.en-us.md [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+**Table of Contents**
+
+{{< toc >}}
+
+## Act
+
+The [nektos/act](https://github.com/nektos/act) project is an excellent tool that allows you to run your GitHub Actions locally.
+We were inspired by this and wondered if it would be possible to run actions for Gitea.
+
+However, while [nektos/act](https://github.com/nektos/act) is designed as a command line tool, what we actually need is a Go library with adjustments for Gitea.
```
Comment (silverwind) [inline docs/content/doc/actions/design.en-us.md]:
```suggestion
However, while [nektos/act](https://github.com/nektos/act) is designed as a command line tool, what we actually needed was a Go library with modifications specifically for Gitea.
```

File: docs/content/doc/actions/faq.en-us.md [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+You can use `github.xyz` and Gitea will work fine.
+As mentioned, Gitea Actions is designed to be compatible with GitHub Actions.
+However, we recommend using `gitea.xyz` in case Gitea adds something that GitHub does not have to avoid different kinds of secrets in your workflow file (and because you are using this workflow on Gitea, not GitHub).
+Still, this is completely optional since both options have the same effect at the moment.
+
+## Is it possible to register runners for a specific user (not organization)?
+
+Not yet.
+It is technically possible to implement, but we need to discuss whether it is necessary.
+
```
Comment (silverwind) [inline docs/content/doc/actions/faq.en-us.md:49]:
```suggestion
```

File: docs/content/doc/actions/faq.en-us.md [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+```
+
+## Why choose GitHub Actions? Why not something compatible with GitLab CI/CD?
+
+[@lunny](https://gitea.com/lunny) has explained this in the [issue to implement actions](https://github.com/go-gitea/gitea/issues/13539).
+Furthermore, Actions is not only a CI/CD system but also an automation tool.
+
+There have also been numerous [marketplace actions](https://github.com/marketplace?type=actions) implemented in the open-source world.
+It is exciting to be able to reuse them.
+
```
Comment (silverwind) [inline docs/content/doc/actions/faq.en-us.md:121]:
```suggestion
```

File: docs/content/doc/actions/faq.en-us.md [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+## Will there be more implementations for Gitea Actions runner?
+
+Although we would like to provide more options, our limited manpower means that act runner will be the only officially supported runner.
+However, both Gitea and act runner are completely open source, so anyone can create a new/better implementation.
+We support your choice, no matter how you decide.
+In case you fork act runner to create your own version: Please contribute the changes back if you can and if you think your changes will help others as well.
+
+
+
+
```
Comment (silverwind) [inline docs/content/doc/actions/faq.en-us.md]:
```suggestion
```

File: docs/content/doc/actions/quickstart.en-us.md [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+You can obtain tokens from `<your_gitea.com>/admin/runners`.
+
+![register runner](/images/usage/actions/register-runner.png)
+
+After registering, a new file named `.runner` will appear in the current directory.
+This file stores the registration information.
+Please do not edit it manually.
+If this file is missing or corrupted, you can simply remove it and register again.
+
+Finally, it's time to start the runner.
```
Comment (silverwind) [inline docs/content/doc/actions/quickstart.en-us.md]:
```suggestion
Finally, it's time to start the runner:
```

File: docs/content/doc/actions/quickstart.en-us.md [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+To enable it, go to the settings page of your repository like `your_gitea.com/<owner>/repo/settings` and enable `Enable Repository Actions`.
+
+![enable actions](/images/usage/actions/enable-actions.png)
+
+The next steps may be rather complicated.
+You will need to study [the workflow syntax](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions) for Actions and write the workflow files you want.
+
+However, we can just start from a simple demo:
+
+
```
Comment (silverwind) [inline docs/content/doc/actions/quickstart.en-us.md]:
```suggestion
```

File: docs/content/doc/actions/quickstart.en-us.md [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+```yaml
+name: Gitea Actions Demo
+run-name: ${{ gitea.actor }} is testing out Gitea Actions 🚀
+on: [push]
+jobs:
```
Comment (silverwind) [inline docs/content/doc/actions/quickstart.en-us.md:108]:
```suggestion

jobs:
```

Personal preference.

File: docs/content/doc/actions/quickstart.en-us.md [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+If the charset is not `utf8mb4`, errors will occur, such as `Error 1366 (HY000): Incorrect string value: '\\xF0\\x9F\\x8E\\x89 T...' for column 'name' at row 1`.
+See [Database Preparation]( {{ < relref "doc/installation/database-preparation.en-us.md#mysql" > }}) for more information.
+
+Alternatively, you can use this demo without emojis:
+
+```yaml
+name: Gitea Actions Demo
+run-name: ${{ gitea.actor }} is testing out Gitea Actions
+on: [push]
+jobs:
```
Comment (silverwind) [inline docs/content/doc/actions/quickstart.en-us.md]:
```suggestion

jobs:
```

File: docs/content/doc/actions/quickstart.en-us.md [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+You can upload it as a file with the extension `.yaml` in the directory `.gitea/workflows/` of the repository, for example `.gitea/workflows/demo.yaml`.
+You might notice that this is fairly similar from the [Quickstart for GitHub Actions](https://docs.github.com/en/actions/quickstart).
+That is because  Gitea Actions is designed to be compatible with GitHub Actions wherever possible.
+
+Be careful, the demo file contains some emojis.
+Please make sure your database supports them, especially when using MySQL.
+If the charset is not `utf8mb4`, errors will occur, such as `Error 1366 (HY000): Incorrect string value: '\\xF0\\x9F\\x8E\\x89 T...' for column 'name' at row 1`.
+See [Database Preparation]( {{ < relref "doc/installation/database-preparation.en-us.md#mysql" > }}) for more information.
+
+Alternatively, you can use this demo without emojis:
```
Comment (silverwind) [inline docs/content/doc/actions/quickstart.en-us.md]: Is it really necessary to duplicate just for the emoji removal?

File: docs/content/doc/actions/quickstart.en-us.md [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+          ls ${{ gitea.workspace }}
+      - run: echo "This job's status is ${{ gitea.status }}."
+```
+
+The line `on: [push]` indicates that the workflow will be triggered when you push commits to this repository.
+However, when you upload the YAML file, it also pushes a commit, so you should see a new task in the Actions tab.
+
+![view job](/images/usage/actions/view-job.png)
+
+
```
Comment (silverwind) [inline docs/content/doc/actions/quickstart.en-us.md]:
```suggestion
```

File: docs/content/doc/actions/act-runner.en-us.md (hunk 11 above)
Comment (techknowlogick) [reply docs/content/doc/actions/act-runner.en-us.md]: Likely "if the default image is insufficient for your needs" is more appropriate

File: docs/content/doc/actions/act-runner.en-us.md (hunk 10 above)
Comment (lunny) [reply docs/content/doc/actions/act-runner.en-us.md:183]: They are `act` default images, but of course we can change them?

File: docs/content/doc/actions/comparison.en-us.md [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+
+Even though Gitea Actions is designed to be compatible with GitHub Actions, there are some differences between them.
+
+**Table of Contents**
+
+{{< toc >}}
+
+## Additional features
+
+### Absolute action URLs
```
Comment (lunny) [inline docs/content/doc/actions/comparison.en-us.md:26]: I think configurable actions site should also be mentioned here. That means, the default site is not github.com and configurable.

File: docs/content/doc/actions/act-runner.en-us.md (hunk 10 above)
Comment (silverwind) [reply docs/content/doc/actions/act-runner.en-us.md:183]: I'm not sure the alternatives. Languages like node or golang all use debian at their base image, but I guess the community prefers ubuntu for compat with github's fat images that contain like 20+ different languages, the label needs to be made compatible.

**Comment by silverwind** [comment]
Some markdownlint issues:

https://github.com/go-gitea/gitea/actions/runs/4853405379/jobs/8649532073?pr=24405

File: docs/content/doc/actions/act-runner.en-us.md (hunk 11 above)
Comment (wolfogre) [reply docs/content/doc/actions/act-runner.en-us.md]: What I tried to say was that the default image is small but it lacks a lot of tools, so it's "poor".

I think techknowlogick's suggestion is great.

```suggestion
If the default image is insufficient for your needs, and you have enough disk space to use a better and bigger one, you can change it to `ubuntu-22.04:docker://<the image you like>`.
```

File: docs/content/doc/actions/quickstart.en-us.md (hunk 28 above)
Comment (wolfogre) [reply docs/content/doc/actions/quickstart.en-us.md]: Removed at 3c9a0e2ff9

File: docs/content/doc/actions/comparison.en-us.md (hunk 30 above)
Comment (wolfogre) [reply docs/content/doc/actions/comparison.en-us.md:26]: Added in "Different behavior"

See 6024cfc04d

**Review by lunny — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**